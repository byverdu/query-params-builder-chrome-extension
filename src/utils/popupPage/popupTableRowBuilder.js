/**
 * @typedef {Object} ExtendedProps
 * @property {boolean} withExtraThead
 * @typedef {import('../../types/index.js').ExtensionOptions & ExtendedProps} Props
 */

/**
 * @param {Props} param
 * @returns string
 */
export function popupTableRowBuilder({
  id,
  bundleName,
  urlParamKey,
  checked,
  urlParamValue,
  canDeleteFromPopup,
  withExtraThead,
}) {
  return `
  <tr>
    <td class="td-checkbox">
      <input
        class="form-check-input me-1"
        type="checkbox"
        value="${urlParamKey}"
        data-url-param-key="${urlParamKey}"
        data-bundle-name="${bundleName}"
        id="${id}"
        ${(checked && 'checked') || ''}
        data-can-delete-from-popup="${canDeleteFromPopup}"
      >
      <label class="form-check-label" for="${id}">${bundleName}</label>
    </td>
    <td>
      <input
        type="text"
        class="form-control"
        data-id="${id}"
        placeholder="${urlParamKey} value"
        value="${(urlParamValue && urlParamValue) || ''}"
      />
    </td>
    ${
      (canDeleteFromPopup &&
        '<td><button class="btn btn-danger delete-new-item">-</button></td>') ||
      (withExtraThead && '<td></td>') ||
      ''
    }
  </tr>
`;
}
