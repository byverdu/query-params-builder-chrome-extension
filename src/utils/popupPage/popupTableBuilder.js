import { popupTableRowBuilder } from './popupTableRowBuilder.js';

/**
 *
 * @param {HTMLTableElement} table
 */
function appendExtraTableHeader(table) {
  const tableHeader = table.querySelector('thead tr');

  if (!tableHeader) {
    throw new Error('No thead tr selector present on the document');
  }

  tableHeader.insertAdjacentHTML(
    'beforeend',
    '<th id="withExtraThead" scope="col">Actions</th>'
  );
}

/**
 * @param {import('../../types/index.js').ExtensionOptions[]} optionsToBuild
 */
export function popupTableBuilder(optionsToBuild) {
  /**
   * @type HTMLTableElement | null
   */
  const table = document.querySelector('.table');

  if (!table) {
    throw new Error('No .table selector present on the document');
  }

  const tbody = table.querySelector('tbody');
  let tableRows = '';
  const withExtraThead = optionsToBuild.some(
    option => option.canDeleteFromPopup
  );

  if (withExtraThead && !document.getElementById('withExtraThead')) {
    appendExtraTableHeader(table);
  }

  for (const {
    id,
    bundleName,
    urlParamKey,
    urlParamValue,
    checked,
    canDeleteFromPopup,
  } of optionsToBuild) {
    tableRows += popupTableRowBuilder({
      id,
      bundleName,
      urlParamKey,
      checked,
      urlParamValue,
      canDeleteFromPopup,
      withExtraThead,
    });
  }

  if (tbody) {
    tbody.insertAdjacentHTML('beforeend', tableRows);
  }

  if (optionsToBuild.length === 1 && withExtraThead) {
    tbody.querySelectorAll('tr').forEach(elem => {
      const withoutDeleteBtn = elem.querySelectorAll('td').length === 2;
      if (withoutDeleteBtn) {
        elem.insertAdjacentHTML('beforeend', '<td></td>');
      }
    });
  }
}
