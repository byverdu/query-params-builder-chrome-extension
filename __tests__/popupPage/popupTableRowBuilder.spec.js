import { popupTableRowBuilder } from '../../src/utils/popupPage/popupTableRowBuilder';
import { propsBuilder } from './helpers';

describe('popupTableRowBuilder', () => {
  it('should be defined', () => {
    expect(popupTableRowBuilder).toBeInstanceOf(Function);
  });

  it('should render the <tr>', () => {
    const props = propsBuilder();
    const markup = popupTableRowBuilder(props);

    expect(markup).toMatchSnapshot();
  });

  it('should render different <tr> if items can be deleted from popup', () => {
    const props = propsBuilder({
      canDeleteFromPopup: true,
      withExtraThead: true,
    });
    const markup = popupTableRowBuilder(props);

    expect(markup).toMatchSnapshot();
  });

  it('should render different <tr> if items are not checked', () => {
    const props = propsBuilder({
      checked: false,
    });
    const markup = popupTableRowBuilder(props);

    expect(markup).toMatchSnapshot();
  });

  it('should render different <tr> if urlParamValue is not provided', () => {
    const props = propsBuilder({
      urlParamValue: undefined,
    });
    const markup = popupTableRowBuilder(props);

    expect(markup).toMatchSnapshot();
  });
});
