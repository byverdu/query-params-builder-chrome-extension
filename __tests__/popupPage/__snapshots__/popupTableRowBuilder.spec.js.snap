// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popupTableRowBuilder should render different <tr> if items are not checked 1`] = `
"
  <tr>
    <td class="td-checkbox">
      <input
        class="form-check-input me-1"
        type="checkbox"
        value="apiKey"
        data-url-param-key="apiKey"
        data-bundle-name="api key"
        id="123"
        
        data-can-delete-from-popup="false"
      >
      <label class="form-check-label" for="123">api key</label>
    </td>
    <td>
      <input
        type="text"
        class="form-control"
        data-id="123"
        placeholder="apiKey value"
        value="isASecret"
      />
    </td>
    
  </tr>
"
`;

exports[`popupTableRowBuilder should render different <tr> if items can be deleted from popup 1`] = `
"
  <tr>
    <td class="td-checkbox">
      <input
        class="form-check-input me-1"
        type="checkbox"
        value="apiKey"
        data-url-param-key="apiKey"
        data-bundle-name="api key"
        id="123"
        checked
        data-can-delete-from-popup="true"
      >
      <label class="form-check-label" for="123">api key</label>
    </td>
    <td>
      <input
        type="text"
        class="form-control"
        data-id="123"
        placeholder="apiKey value"
        value="isASecret"
      />
    </td>
    <td><button class="btn btn-danger delete-new-item">-</button></td>
  </tr>
"
`;

exports[`popupTableRowBuilder should render different <tr> if urlParamValue is not provided 1`] = `
"
  <tr>
    <td class="td-checkbox">
      <input
        class="form-check-input me-1"
        type="checkbox"
        value="apiKey"
        data-url-param-key="apiKey"
        data-bundle-name="api key"
        id="123"
        checked
        data-can-delete-from-popup="false"
      >
      <label class="form-check-label" for="123">api key</label>
    </td>
    <td>
      <input
        type="text"
        class="form-control"
        data-id="123"
        placeholder="apiKey value"
        value=""
      />
    </td>
    
  </tr>
"
`;

exports[`popupTableRowBuilder should render the <tr> 1`] = `
"
  <tr>
    <td class="td-checkbox">
      <input
        class="form-check-input me-1"
        type="checkbox"
        value="apiKey"
        data-url-param-key="apiKey"
        data-bundle-name="api key"
        id="123"
        checked
        data-can-delete-from-popup="false"
      >
      <label class="form-check-label" for="123">api key</label>
    </td>
    <td>
      <input
        type="text"
        class="form-control"
        data-id="123"
        placeholder="apiKey value"
        value="isASecret"
      />
    </td>
    
  </tr>
"
`;
